sycl/doc/SYCLGraphUsageGuide.md

```diff
@@ -0,0 +1,493 @@
+    // Do some host work here to prepare for the kernel to be executed
+    do_some_host_work();
+    CGH.parallel_for(...);
+});
+```
+
+In normal SYCL usage `do_some_host_work()` will be evaluated during the call to
+submit. With SYCL-Graph this code will be evaluated once during the call to
+`submit()` or `command_graph::add()` but will not be evaluated on subsequent
+runs of the graph which may lead to incorrect behaviour. This code should be
+placed into a `host-task` node which the kernel (or other operation) depends on,
+like so:
+
+```c++
+sycl::event HostWorkEvent = Queue.submit([&](sycl::handler& CGH){
+    CGH.host_task([=](){
+        do_some_host_work();
+    });
+});
+
+Queue.submit([&](sycl::handler& CGH){
+    CGH.depends_on(HostWorkEvent);
+    CGH.parallel_for(...);
+});
+```
+
+### Graph Execution Concurrency
+
+Normally the SYCL runtime will prevent multiple submissions of a given
+`command_graph` from executingly concurrently to prevent data races when
+accessing the same memory resources. However, it is quite common to wish to
+execute the same graph on different sets of data concurrently. This can be
+accomplished by using graph update functionality, but this must also be
+serialized with a host synchronization. Consider this example (it uses whole
+graph update for simplicity):
+
+```c++
+sycl::command_graph<graph_state::executable> ExecutableGraph
+   = ModifiableGraph.finalize();
+
+Queue.ext_oneapi_graph(ExecutableGraph);
+// Updating the graph here to use new memory, this forces a host synchronization
+ExecutableGraph.update(ModifiableGraphWithNewParams);
+
+// Re-execute the update graph
+Queue.ext_oneapi_graph(ExecutableGraph);
+
+```
+
+If your sets of inputs are known (such as a double-buffering type scenario), you
+can create multiple executable graphs up front, avoiding the need for expensive
+host sync between executions (assuming there are no implicit dependencies from
+accessing the same `sycl::buffer`) for updating, and potentially increasing
+device occupancy (if the device supports it). Modifying the example from above:
+
+```c++
+sycl::command_graph<graph_state::executable> ExecutableGraph
+   = ModifiableGraph.finalize();
+sycl::command_graph<graph_state::executable> ExecutableGraphOtherParams
+   = ModifiableGraphWithNewParams.finalize();
+
+Queue.ext_oneapi_graph(ExecutableGraph);
+// We can now execute the second graph with no implicit dependency
+Queue.ext_oneapi_graph(ExecutableGraphOtherParams);
+
+// Repeatedly executing the graphs will create dependencies on their individual
+// preceeding executions, but not on each other.
+Queue.ext_oneapi_graph(ExecutableGraph);
+Queue.ext_oneapi_graph(ExecutableGraphOtherParams);
+
+```
+
+If the amount of different inputs is not always known up front a similar
+strategy could be employed but instead using a pool of graphs to hide the
+potential host-synchronization caused when updating and increase device
+occupancy.
+
+
+### Recording Library Calls
+
```

Comment:
We have to think about how to best collect limitations with current oneAPI library calls. Which are a moving target, as we incrementally improve the support in future versions. I'm missing a note here what is supported.

It would be also great to have a concrete example here, maybe DNN or MKL, that works.

Reply:
I think because it is still very early days in terms of library support we should be cautious about including anything definitive here. I included this specific scenario because there was an open PR to add a note about this to the spec, but felt it would be better included here (when we got around to actually creating the document).

Reply:
I've included a disclaimer about library compatibility now prefacing the subsequent example.